src/XR/features/WebXRNearInteraction.ts

```diff
@@ -0,0 +1,590 @@
+                let hoverGrabInfo = accuratePickInfo(originalSceneHoverGrab, utilitySceneHoverGrab);
+                if ((hoverPickInfo && hoverPickInfo.hit) || (hoverGrabInfo && hoverGrabInfo.hit)) {
+                    // turn off far interaction if in the hover range
+                    if (this._farInteractionFeature) {
+                        this._farInteractionFeature.detach();
+                        hoverRange = true;
+                    }
+                }
+                pick = populateNearInteractionInfo(hoverPickInfo);
+            }
+
+            // near interaction pick
+            if (controllerData.pickIndexMeshTip) {
+                let utilitySceneNearPick = null;
+                if (this._options.useUtilityLayer && this._utilityLayerScene) {
+                    utilitySceneNearPick = this._pickWithMesh(controllerData.pickIndexMeshTip, this._utilityLayerScene, false, this.nearPickPredicate);
+                }
+                let originalSceneNearPick = this._pickWithMesh(controllerData.pickIndexMeshTip, this._scene, false, this.nearPickPredicate);
+                let pickInfo = accuratePickInfo(originalSceneNearPick, utilitySceneNearPick);
+                pick = populateNearInteractionInfo(pickInfo);
+            }
+
+            // near interaction grab
+            if (controllerData.grabRay) {
+                let utilitySceneNearGrab = null;
+                if (this._utilityLayerScene) {
+                    utilitySceneNearGrab = this._utilityLayerScene.pickWithRay(controllerData.grabRay, this.nearGrabPredicate);
+                }
+                let originalSceneNearGrab = this._scene.pickWithRay(controllerData.grabRay, this.nearGrabPredicate);
+                pick = accuratePickInfo(originalSceneNearGrab, utilitySceneNearGrab);
+                if (pick && pick.pickedPoint && pick.hit) {
+                    controllerData.nearGrab = true;
+                    controllerData.nearInteraction = true;
+                    pick.ray = controllerData.grabRay;
+                }
+            }
+
+            // Turn on far interaction if near interaction is unsuccessful
+            if (!controllerData.nearInteraction && !hoverRange && !controllerData.nearGrabInProcess) {
+                if (this._farInteractionFeature) {
+                    this._farInteractionFeature.attach();
+                }
+            }
+            controllerData.pick = pick;
+
+            if (pick && pick.pickedPoint && pick.hit) {
+                controllerData.meshUnderPointer = pick.pickedMesh;
+            } else {
+                controllerData.meshUnderPointer = null;
+            }
+        });
+    }
+
+    private get _utilityLayerScene() {
+        return this._options.customUtilityLayerScene || UtilityLayerRenderer.DefaultUtilityLayer.utilityLayerScene;
+    }
+
+    private _attachTrackedPointerRayMode(xrController: WebXRInputSource) {

```

Comment:
This function can be reduced and renamed, since it's no longer attaching a tracked pointer ray.

controllerData.SqueezeComponent is what you want for hands, as the grasp event triggers consistently, which the primary trigger triggers only on certain hand orientations (for holoLens, at least). It doesn't make sense to have active listeners to both though, as that should be doubling the input events.

controllerData.SelectComponent would be what we want for other, physical motion controllers, except that the spheres we track hit-testing with are tracking specific hand joints that don't exist on non-hand controllers. So if other hands (such as Oculus) do not send grasp, you would need to listen to this event, otherwise just listening to the squeeze should be more than enough.

Also the event listeners at the bottom can probably be removed, as from what I remember Raanan said they were added to get hand selection working prior to hands gaining the gamepad object.

I remember suggesting some of these extra parts could be kept in to make testing easier while the API has changed, let me know if I mentioned a different reason to keep them that I've now forgotten.

Reply:
Sure, I agree with renaming. Initially i did not have the select component added but added it later for supporting oculus specifically after Raanan got to test with it. @RaananW  Could you please confirm if I can remove event listeners?

Reply:
Ah yeah, that was the part I was unsure about. I'm still worried about repeated events on HoloLens though - is it possible to do something where if "grasp" exists listen to squeeze, otherwise listen to the select action?

Reply:
Isn't that the case right now? This is what I do in the pointer selection - if no gamepad is available, use the events.

Reply:
I don't see a case where events will be repeated. And hololens will never go into the event listener registration loop